### ISSUE TITLE
Nix multi-user install doesn't support non-login non-interactive shells on Ubuntu (which breaks distributed builds)

### ISSUE THREAD

**Issue by djahandarie** [NixOS/nix#2587]
Using the multi-user install on an Ubuntu machine `nixhost` followed by running commands of the form `ssh nixhost nix-<whatever>` result in the following error:

```
cannot build on 'ssh://XXXX': cannot connect to 'XXXX': bash: nix-store: command not found
```

This means that the host cannot be used as a `buildMachine` in a distributed build setup.

Investigation
---------------

A basic multi-user-install on Ubuntu modifies `/etc/profile.d/nix.sh`, because `/etc/bashrc` doesn't exist, only `/etc/bash.bashrc` does.

`/etc/profile`, which calls `/etc/profile.d/nix.sh`, is only called for full login shells.

This means the code is only called for full login shells (interactive or not).

Thoughts
---------------

If we also appended the nix sourcing to `/etc/bash.bashrc` (by modifying [PROFILE_TARGETS in  install-multi-user.sh](https://github.com/NixOS/nix/blob/6ba1726eeb58cd8e7d69e2a090346f26fdf634a5/scripts/install-multi-user.sh)), that would at least result in the code running on non-login *interactive* shells, but that still would not result in the code being run on non-login *non-interactive* shells, due to a check at the top of `/etc/bash.bashrc`

[An askubuntu answer](https://askubuntu.com/a/247769) suggests *prepending* to `/etc/bash.bashrc` if you want to execute commands on non-login non-interactive shells.

Questions
--------------

1. Do we want nix to be usable from non-login non-interactive shells by default?

2. Do we want to accomplish the above by prepending to `/etc/bash.bashrc`?

3. If we don't, should we consider changing the distributed builder to use full login shells (like [this superuser Q&A documents](https://superuser.com/questions/306530/run-remote-ssh-command-with-full-login-shell))?

cc @grahamc

**Comment by knedlsepp** [comment]
Having the same issue on CentOS. It did however work for the root user out of the box. The issue was that `~root/.bashrc` contained the line `source /etc/bashrc`. So my personal solution is just to put the following in my personal bashrc on that server:
```bash
if [ -f /etc/bashrc ]; then
        . /etc/bashrc
fi
```

**Comment by mschwaig** [comment]
I ran into this on an Ubuntu 18.04 box with a Multi-User Install of Nix when I tried to use `nix-copy-closure` or `nix copy` to copy  some store paths over there.

I solved it by pre-pending the following to `/etc/bash.bashrc`
```
if [ -e /etc/profile.d/nix.sh ]; then
  . /etc/profile.d/nix.sh
fi
```
I guess this is kind of what the installer would append to other files if I understand correctly.

I think it would be great if this worked out of the box assuming there are no good reasons (engineering?/security?/user-expectations?) why it should not.

Are others also still running into this issue? Can someone think of a reason why it should not work? 🤔

**Comment by bjornfor** [comment]
> Are others also still running into this issue? Can someone think of a reason why it should not work?

Yes, I also work around this issue by inserting that code snippet. It feels a bit hacky, I'd much rather drop something in a `/etc/foo.d` directory. But if having to chose I'd definitely chose having a working `nix-copy-closure` out of the box over keeping `/etc/bash.bashrc` untouched :-)

**Comment by RyanGibb** [comment]
> I ran into this on an Ubuntu 18.04 box with a Multi-User Install of Nix when I tried to use `nix-copy-closure` or `nix copy` to copy some store paths over there.
> 
> I solved it by pre-pending the following to `/etc/bash.bashrc`
> 
> ```
> if [ -e /etc/profile.d/nix.sh ]; then
>   . /etc/profile.d/nix.sh
> fi
> ```
> 
> I guess this is kind of what the installer would append to other files if I understand correctly.
> 
> I think it would be great if this worked out of the box assuming there are no good reasons (engineering?/security?/user-expectations?) why it should not.
> 
> Are others also still running into this issue? Can someone think of a reason why it should not work? 🤔

Just ran into this and this fix worked.
